eth/tracers/api_test.go

```diff
@@ -637,6 +745,318 @@ func TestTraceBlock(t *testing.T) {
+		Alloc: types.GenesisAlloc{
+			accounts[0].addr: {Balance: big.NewInt(params.Ether)},
+		},
+	}
+	signer := types.HomesteadSigner{}
+	var badBlockHash common.Hash
+
+	backend := newTestBackend(t, 2, genesis, func(i int, b *core.BlockGen) {
+		tx, _ := types.SignTx(types.NewTx(&types.LegacyTx{
+			Nonce:    uint64(i),
+			To:       &accounts[1].addr,
+			Value:    big.NewInt(1000),
+			Gas:      params.TxGas,
+			GasPrice: b.BaseFee(),
+			Data:     nil,
+		}), signer, accounts[0].key)
+		b.AddTx(tx)
+	})
+	defer backend.teardown()
+
+	// Create a bad block by corrupting the first block after genesis
+	badBlock := backend.chain.GetBlockByNumber(1)
+	if badBlock == nil {
+		t.Fatalf("failed to fetch block 1")
+	}
+	badBlock.Header().ParentHash = common.HexToHash("0xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef") // Simulate corruption
+	badBlockHash = badBlock.Hash()
+	rawdb.WriteBadBlock(backend.chaindb, badBlock)
+
+	// Test cases
+	api := NewAPI(backend)
+	var testCases = []struct {
+		name       string
+		hash       common.Hash
+		config     *StdTraceConfig
+		expectErr  bool
+		expectMsg  string
+		validateFn func(files []string)
+	}{
+		{
+			name:      "Valid bad block",
+			hash:      badBlockHash,
+			config:    nil, // Use default StdTraceConfig
+			expectErr: false,
+			validateFn: func(files []string) {
+				if len(files) == 0 {
+					t.Fatalf("expected at least one output file, got none")
+				}
+				for _, file := range files {
+					if _, err := os.Stat(file); os.IsNotExist(err) {
+						t.Fatalf("expected output file %s to exist", file)
+					}
+					// Clean up the file after test
+					os.Remove(file)
+				}
+			},
+		},
+		{
+			name:      "Non-existent block",
+			hash:      common.HexToHash("0xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef"),
+			config:    nil,
+			expectErr: true,
+			expectMsg: fmt.Sprintf("block %s not found", common.HexToHash("0xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef")),
+		},
+	}
+
+	// Execute test cases
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+			files, err := api.StandardTraceBadBlockToFile(ctx, tc.hash, tc.config)
+
+			if tc.expectErr {
+				if err == nil || !strings.Contains(err.Error(), tc.expectMsg) {
+					t.Fatalf("expected error containing '%s', got: %v", tc.expectMsg, err)
+				}
+				return
+			}
+
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+
+			// Validate the output files
+			if tc.validateFn != nil {
+				tc.validateFn(files)
+			}
+		})
+	}
+}
+
+func TestIntermediateRoots(t *testing.T) {
+	t.Parallel()
+
+	// Setup a test backend with valid transactions
+	accounts := newAccounts(2)
+	genesis := &core.Genesis{
+		Config: params.TestChainConfig,
+		Alloc: types.GenesisAlloc{
+			accounts[0].addr: {Balance: big.NewInt(params.Ether)},
+		},
+	}
+	signer := types.HomesteadSigner{}
+	var badBlockHash, genesisHash common.Hash
+
+	backend := newTestBackend(t, 2, genesis, func(i int, b *core.BlockGen) {
```

Comment:
Specifically when it comes to intermediate roots it would be good to modify the state apart from only the transfer. e.g. some SSTORE as part of a tx.